objc2/src/foundation/mutable_set.rs

```diff
@@ -0,0 +1,362 @@
+        /// ];
+        /// let set = NSMutableSet::from_vec(strs);
+        /// let vec = NSMutableSet::into_vec(set);
+        /// assert_eq!(vec.len(), 3);
+        /// ```
+        pub fn into_vec(set: Id<Self, Owned>) -> Vec<Id<T, O>> {
+            set.into_iter()
+                .map(|obj| unsafe { Id::retain(obj as *const T as *mut T).unwrap_unchecked() })
+                .collect()
+        }
+    }
+
+    unsafe impl<T: Message> NSMutableSet<T, Shared> {
+        /// Creates an [`NSMutableSet`] from a slice.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use objc2::foundation::{NSMutableSet, NSString};
+        /// # #[cfg(feature = "gnustep-1-7")]
+        /// # unsafe { objc2::__gnustep_hack::get_class_to_force_linkage() };
+        ///
+        /// let strs = ["one", "two", "three"].map(NSString::from_str);
+        /// let set = NSMutableSet::from_slice(&strs);
+        /// ```
+        pub fn from_slice(slice: &[Id<T, Shared>]) -> Id<Self, Owned> {
+            // SAFETY:
+            // Taking `&T` would not be sound, since the `&T` could come from
+            // an `Id<T, Owned>` that would now no longer be owned!
+            //
+            // We always return `Id<NSMutableSet<T, Shared>, Owned>` because
+            // the elements are shared and mutable sets are always unique.
+            unsafe { with_objects(Self::class(), slice.as_slice_ref()) }
+        }
+    }
+
+    // We're explicit about `T` being `PartialEq` for these methods because the
+    // set compares the input value with elements in the set
+    // For comparison: Rust's HashSet requires similar methods to be `Hash` + `Eq`
+    unsafe impl<T: Message + PartialEq, O: Ownership> NSMutableSet<T, O> {
+        /// Adds a value to the set. Returns whether the value was
+        /// newly inserted.
+        ///
+        /// # Examples
+        ///
+        /// ```
+        /// use objc2::foundation::{NSMutableSet, NSString};
+        /// # #[cfg(feature = "gnustep-1-7")]
+        /// # unsafe { objc2::__gnustep_hack::get_class_to_force_linkage() };
+        ///
+        /// let mut set = NSMutableSet::new();
+        ///
+        /// assert_eq!(set.insert(NSString::from_str("one")), true);
+        /// assert_eq!(set.insert(NSString::from_str("one")), false);
+        /// assert_eq!(set.len(), 1);
+        /// ```
+        #[doc(alias = "addObject:")]
+        pub fn insert(&mut self, value: Id<T, O>) -> bool {
+            // SAFETY:
+            // We take `Id<T, O>` instead of `&T` because `&T` could be a
+            // reference to an owned object which would cause us to have a copy
+            // of an owned object in our set. By taking `Id<T, O>`, we force the
+            // caller to transfer ownership of the value to us, making it safe
+            // to insert the owned object into the set.
+            let contains_value = self.contains(&value);
+            unsafe { msg_send![self, addObject: &*value] }
```

Comment:
It surprised me that this line works, but I guess it does because of some inference rules regarding block statements. I would like to avoid it though, because it seems brittle; for comparison, the following wouldn't work:
```rust
unsafe {
    let contains_value = self.contains(&value);
    msg_send![self, addObject: &*value] // Wouldn't work neither with or without trailing semicolon
    !contains_value
}
```

So please use either `let _: () = ...` or `#[sel(addObject:)] fn add_object(&mut self, &T);`.

Similar for `remove` below

Reply:
I ended up going with the second option of `#[sel(addObject:)] fn add_object(&mut self, &T);` and `#[sel(removeObject:)] fn remove_object(&mut self, &T);`